code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,274 @@
+	for(var/client/L in GLOB.clients)
+		if(L.inactivity >= 6000)
+			continue
+		addtimer(L.update_exp_list(mins, ann),10)
+		CHECK_TICK
+
+//Manual incrementing/updating
+/client/proc/update_exp_client(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+	if(!SSdbcore.Connect())
+		return -1
+	var/datum/DBQuery/exp_read = SSdbcore.NewQuery("SELECT job, minutes FROM [format_table_name("role_time")] WHERE ckey = '[sanitizeSQL(ckey)]'")
+	if(!exp_read.Execute())
+		var/err = exp_read.ErrorMsg()
+		log_game("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		message_admins("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		return
+	var/list/play_records = list()
+	while(exp_read.NextRow())
+		play_records[exp_read.item[1]] = text2num(exp_read.item[2])
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
+			if(GLOB.exp_jobsmap[category]["titles"])
+				if(mob.mind.assigned_role in GLOB.exp_jobsmap[category]["titles"])
+					play_records[category] += minutes
+					if(announce_changes)
+						to_chat(mob,"<span class='notice'>You got: [minutes] [category] EXP!")
+		if(mob.mind.special_role)
+			play_records[EXP_TYPE_SPECIAL] += minutes
+			if(announce_changes)
+				to_chat(mob,"<span class='notice'>You got: [minutes] Special EXP!")
+	else if(isobserver(mob))
+		play_records[EXP_TYPE_GHOST] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Ghost EXP!")
+	else if(minutes)	//Let "refresh" checks go through
+		return
+	prefs.exp = play_records
+
+	for(var/jtype in play_records)
+		var jobname = jtype
+		var time = play_records[jtype]
+		var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(ckey)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = '[sanitizeSQL(time)]'")
+		if(!update_query.Execute())
+			var/err = update_query.ErrorMsg()
+			log_game("SQL ERROR during exp_update_client update. Error : \[[err]\]\n")
+			message_admins("SQL ERROR during exp_update_client update. Error : \[[err]\]\n")
+			return
+
+
+
+/client/proc/update_exp_list(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+
+	var/list/play_records = prefs.exp
+	var/list/old_records = prefs.exp.Copy()
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
```

Comment:
And with primary key indexing on the two columns, ckey and job, it will be a very fast lookup.

Lets say that all 50,000 players in the player table have a record, lets say each player has 30 jobs recorded.

50000 players + 30 job = 1,500,000 rows

Mysql uses a sorted binary tree for primary key indexes, to do the query to find every job row for a ckey, since the ckey is the first column in the pimary key index, it will do a binary search for the first matching ckey row in the index then it grab all the rows that follow until it gets to the end of that ckey.

A binary search is O(log2(n)), so at 1,500,000 rows, it will find the first one after 21 lookups.

So thats 2,520 lookups **total**, for all 120 people reconnecting.

In comparison, every time a client connects, a row is inserted in the connection_log database, and this triggers a sorted insert of like 5 columns into indexes, each having to do 23 look ups, so thats 115 lookups per connection, or 13,000 lookups total for 120 players reconnecting at reboot.